ios/CodePush/CodePushUpdateUtils.m

```diff
@@ -256,7 +280,88 @@ + (BOOL)verifyHashForDiffUpdate:(NSString *)finalUpdateFolder
+{
+    publicKeyString = [publicKeyString stringByReplacingOccurrencesOfString:@"-----BEGIN PUBLIC KEY-----\n"
+                                                                 withString:@""];
+    publicKeyString = [publicKeyString stringByReplacingOccurrencesOfString:@"-----END PUBLIC KEY-----"
+                                                                 withString:@""];
+    publicKeyString = [publicKeyString stringByReplacingOccurrencesOfString:@"\n"
+                                                                 withString:@""];
+
+    return publicKeyString;
+}
+
++ (NSString *)getSignatureFor:(NSString *)folderPath
+                        error:(NSError **)error
+{
+    NSString *signatureFilePath = [NSString stringWithFormat:@"%@/%@/%@", folderPath, ManifestFolderPrefix, BundleJWTFile];
+    if ([[NSFileManager defaultManager] fileExistsAtPath:signatureFilePath]) {
+        return [NSString stringWithContentsOfFile:signatureFilePath encoding:NSUTF8StringEncoding error:error];
+    }else{
+        *error = [CodePushErrorUtils errorWithMessage:[NSString stringWithFormat: @"Cannot find signature at %@", signatureFilePath]];
+        return nil;
+    }
+    return nil;
+}
+
++ (NSDictionary *) verifyJWT:(NSString *) signature
+     withPublicKey:(NSString *)publicKey
+             error:(NSError **)error
+{
+    id <JWTAlgorithmDataHolderProtocol> verifyDataHolder = [JWTAlgorithmRSFamilyDataHolder new].keyExtractorType([JWTCryptoKeyExtractor publicKeyWithPEMBase64].type).algorithmName(@"RS256").secret(publicKey);
+    
+    JWTCodingBuilder *verifyBuilder = [JWTDecodingBuilder decodeMessage:signature].addHolder(verifyDataHolder);
+    JWTCodingResultType *verifyResult = verifyBuilder.result;
+    if (verifyResult.successResult) {
+        return verifyResult.successResult.payload;
+    }
+    else {
+        *error = verifyResult.errorResult.error;
+        return nil;
+    }
+}
+
++ (BOOL)verifySignatureFor:(NSString *)folderPath
+             withPublicKey:(NSString *)publicKeyString
+                     error:(NSError **)error
+{
+    NSLog(@"Verifying signature for folder path: %@", folderPath);
+    
+    NSString *publicKey = [self cleanPublicKey: publicKeyString];
+    
+    NSString *signature = [self getSignatureFor: folderPath
+                                          error: error];
+    if(signature == nil) {
+        if(error && *error){
+            CPLog(@"The update could not be verified because no signature was found. %@", *error);
+        }else{
+            CPLog(@"The update could not be verified because no signature was found.");
+        }
+        return false;
+    }
+    
+    NSDictionary *envelopedPayload = [self verifyJWT:signature withPublicKey:publicKey error:error];
+    if(envelopedPayload == nil){
+        CPLog(@"The update could not be verified because it was not signed by a trusted party. %@", *error);
+        return false;
+    }
+    
+    if(![envelopedPayload objectForKey:@"contentHash"]){
```

Comment:
Probably could move `@contentHash` to constants

Reply:
@max-mironov not so important for now, i think. May be later, if payload structure will become more complicated.

Reply:
Agreed!